Sources/StreamChat/WebSocketClient/ConnectionStatus.swift

```diff
@@ -32,17 +32,20 @@ extension ConnectionStatus {
         case .initialized:
             self = .initialized
             
-        case let .disconnected(source):
-            self = .disconnected(error: source.serverError)
-            
-        case .connecting, .waitingForConnectionId, .waitingForReconnect:
+        case .connecting, .waitingForConnectionId:
             self = .connecting
             
         case .connected:
             self = .connected
             
         case .disconnecting:
             self = .disconnecting
+            
+        case let .disconnected(source):
+            let isWaitingForReconnect = webSocketConnectionState.isAutomaticReconnectionEnabled || source.serverError?
+                .isInvalidTokenError == true
+            
+            self = isWaitingForReconnect ? .connecting : .disconnected(error: source.serverError)
```

Comment:
That is not nice at all but didn't find any better way 😕 The thing is we don't have token refreshing logic encapsulated so there's no easy way to move it from `ChatClient` to `ConnectionRecoveryHandler`. That's why in case of an expired token `ConnectionRecoveryHandler ` does nothing (since `isAutomaticReconnectionEnabled` returns false) and just waits the chat client to refetch the token and initiate reconnection by itself.